src/main/java/com/amazon/opendistro/elasticsearch/performanceanalyzer/jvm/ThreadList.java

```diff
@@ -104,37 +111,55 @@ public String toString() {
     }
   }
 
+  /**
+   * This is called from OSMetricsCollector#collectMetrics. So this is not called
+   * in the critical path of ES request handling. Even for the collector thread,
+   * we do a timed wait to acquire this lock and move on if we could not get it.
+   * @return A hashmap of threadId to threadState.
+   */
   public static Map<Long, ThreadState> getNativeTidMap() {
-    synchronized (ThreadList.class) {
-      if (System.currentTimeMillis() > lastRunTime + minRunInterval) {
-        runThreadDump(pid, new String[0]);
+    final int waitTimeOutMillis = 100;
+    try {
+      if (vmAttachLock.tryLock(waitTimeOutMillis, TimeUnit.MILLISECONDS)) {
```

Comment:
I have only one high level question on this. This try lock can let our PA threads to bypass this if they fail to acquire the lock. But what about the ES threads ?  I assume the stucking PA thread can still hold this lock for 10s before it can be released to ES ?

Reply:
This code path is only executed for the collectors. ES threads should just make `getThreadState` and if they could not find the threadId in the Map, they emit a metric and move on. But you are right, the PA thread doing the attach, can get blocked here and therefore we will stop emitting the PA metrics but that should not affect the ES threads anymore.

Reply:
1. It does not matter. a lot of pa metrics are useless if we cannot associate them with os thread ID. 
2. This is a good point. I have removed it.